src/models/PropStore.ts

```diff
@@ -0,0 +1,49 @@
+import { InjectionKey } from 'vue'
+import { CreateRouteOptions, isWithComponent, isWithComponents } from '@/types/createRouteOptions'
+import { ResolvedRoute } from '@/types/resolved'
+import { MaybePromise } from '@/types/utilities'
+
+export const propStoreKey: InjectionKey<PropStore> = Symbol()
+
+type NamedComponentProps = { name: string, props?: (params: Record<string, unknown>) => MaybePromise<Record<string, unknown>> }
+
+export class PropStore {
+  private readonly store = new Map<string, MaybePromise<Record<string, unknown>>>()
+
+  public setProps(route: ResolvedRoute): void {
+    this.store.clear()
+
+    route.matches
+      .flatMap(match => this.getComponentProps(match))
+      .forEach(({ name, props }) => {
+        if (props) {
+          const key = this.getPropKey(route, name)
+          const value = props(route.params)
+
+          this.store.set(key, value)
+        }
+      })
+  }
+
+  public getProps(route: ResolvedRoute, name: string = 'default'): MaybePromise<Record<string, unknown>> | undefined {
```

Comment:
I think name should probably just be required and we can specifically pass in `default` if needed.